### PR TITLE
Add assets directory so that its generated

### LINKED ISSUE #1445: /public/assets is gitignored by default: so builds in the cloud (or off your laptop) don't have /public/assets in them
## Description

If you honor `.gitignore`  to not commit /public/assets, then they don't get shipped into your production containers, and then *local* containers (built from your working buffalo dev environment) will look great, but stuff build outside your dev environment (i.e. directly from source) *won't* have those assets in them.

### Steps to Reproduce the Problem
 
- create a new buffalo project w/ non trivial public/assets
- make sure .gitignore still ignores public/assets
- run the docker build in any environment outside your local laptop (i.e. google cloud build)

You'll ultimately notice that the assets won't be in the same binary, and it will be a megabyte or two smaller then the image you built in your laptop.

## Workaround

Uncomment (in the .gitignore file) the public/assets directory.  This entry is added for you by `buffalo dev` when you created a project.

### PR DESCRIPTION
**Pull request by jayunit100**
fixes #1445  #250 .  This makes it so your cloud build / CI system / whatever will make nice pretty looking buffalo UIs out of the box without needing to load/package your assets in an external place, i.e., its used for cloud build or cloud native style continuous delivery pipelines where you don't have a separate packaging/server step.

Otherwise, the only docker image that will work out of the box will be ones that you build on a machine that already has a populated /public/assets directory (which for most people, is just their laptop).

### CODE REVIEW

**Comment by jayunit100** [comment]
Is there an example i can follow of how to unit test that genny created a file for me (also , does genny actually create the file? I see some reactive style lambda's being passed to it, not sure how the file gets translated into an actuve Mkdir under the hood).

**Comment by markbates** [comment]
Review the unit tests for the code you’re trying to change.

**Comment by jayunit100** [comment]
I guess you mean check around in the APIs ... so far `res.Files` looks promising... ill see what the outputs are...

**Comment by jayunit100** [comment]
@markbates Let me know if this looks about right... thanks for your patience , and thanks again for this lovely framework :) :)

File: genny/build/assets_test.go [hunk 1]
```diff
@@ -37,6 +38,11 @@ func Test_assets(t *testing.T) {
 	for i, c := range res.Commands {
 		r.Equal(cmds[i], strings.Join(c.Args, " "))
 	}
+
+	s, err := res.Find("../public/assets/dummy")
+	if err != nil || s.String() != "placeholder for static builds" {
+		panic(fmt.Sprintf("%v", s.String()))
```
Comment (markbates) [inline genny/build/assets_test.go]: Please follow the rest of the tests and follow the pattern accordingly. We never use `panic`. Thanks.

File: genny/build/assets.go [hunk 2]
```diff
@@ -14,12 +15,15 @@ import (
 )
 
 func assets(opts *Options) (*genny.Generator, error) {
+
 	g := genny.New()
 
+	// Quick way to ensure this directory exists so static content is created.
+	g.File(genny.NewFile("../public/assets/dummy", strings.NewReader("placeholder for static builds")))
```
Comment (markbates) [inline genny/build/assets.go]:
```suggestion
	g.File(genny.NewFile("public/assets/.dummy", strings.NewReader("placeholder for static builds")))
```

File: genny/build/assets_test.go (hunk 1 above)
Comment (jayunit100) [reply genny/build/assets_test.go]: :) yup... forgot I just did this for debugging . I’ll replace with a normal failure .

**Comment by jayunit100** [comment]
@markbates ok, updated to not panic in the middle of the unit test  :)

File: genny/build/assets_test.go [hunk 3]
```diff
@@ -37,6 +37,11 @@ func Test_assets(t *testing.T) {
 	for i, c := range res.Commands {
 		r.Equal(cmds[i], strings.Join(c.Args, " "))
 	}
+
+	s, err := res.Find("public/assets/.dummy")
+	if err != nil || s.String() != "placeholder for static builds" {
```
Comment (markbates) [inline genny/build/assets_test.go]:
```suggestion
	r.NoError(err)
```

File: genny/build/assets_test.go [hunk 4]
```diff
@@ -37,6 +37,11 @@ func Test_assets(t *testing.T) {
...
+		t.Fail()
```
Comment (markbates) [inline genny/build/assets_test.go]:
```suggestion
		r.Equal("placeholder for static builds", f.String())
```

File: genny/build/assets_test.go [hunk 5]
```diff
@@ -37,6 +37,11 @@ func Test_assets(t *testing.T) {
 	for i, c := range res.Commands {
 		r.Equal(cmds[i], strings.Join(c.Args, " "))
 	}
+
+	s, err := res.Find("public/assets/.dummy")
```
Comment (markbates) [inline genny/build/assets_test.go]:
```suggestion
	f, err := res.Find("public/assets/.dummy")
```

**Comment by jayunit100** [comment]
had to switch machines, looking into this now: side note, i had to manually do `go get github.com/stretchr/testify` to get things working.  not sure why.

**Comment by jayunit100** [comment]
travis is unhappy, cant tell why though, its complaining that "public" shouldn't exist.

**Comment by jayunit100** [comment]
pushed a quick fix, should be ok now, will watch travis for updates.

**Comment by jayunit100** [comment]
modified apiapp.json so that it doesnt expect `public` to be missing

**Comment by jayunit100** [comment]
Ok illl look into this now.

**Comment by jayunit100** [comment]
fixed the AsAPI stuff, lets see if travis likes it now.

**Comment by jayunit100** [comment]
yay tests passing nowww
@markbates looks good to you ?

File: buffalo/cmd/filetests/apiapp.json [hunk 6]
```diff
@@ -24,13 +24,7 @@
...
-}, {
-  "path": "templates",
-  "absent": true
-}, {
-  "path": "assets",
-  "absent": true
-}, {
-  "path": "public",
+},{
+  "path": "public/assets/.dummy",
```
Comment (markbates) [inline buffalo/cmd/filetests/apiapp.json]: This shouldn’t exist for API apps.

File: genny/build/assets.go [hunk 7]
```diff
@@ -14,12 +15,15 @@ import (
 )
 
 func assets(opts *Options) (*genny.Generator, error) {
+
 	g := genny.New()
 
+	// Quick way to ensure this directory exists so static content is created.
+	g.File(genny.NewFile("public/assets/.dummy", strings.NewReader("placeholder for static builds")))
```
Comment (markbates) [inline genny/build/assets.go]:
```suggestion
	g.File(genny.NewDir("public/assets"))
```

Comment (markbates) [reply genny/build/assets.go]: Wrap this in an if it check that it should be generated.

File: buffalo/cmd/filetests/apiapp.json (hunk 6 above)
Comment (jayunit100) [reply buffalo/cmd/filetests/apiapp.json]: Absent = true, so that’s what’s being asserted right ?

Comment (markbates) [reply buffalo/cmd/filetests/apiapp.json]: Sorry, on a phone, didn’t see that. You’re correct. :)

**Comment by jayunit100** [comment]
updated again with that `if` block :)

**Comment by jayunit100** [comment]
ok , am actually not doing much on the application side of the house right now, ill try to keep this up to date though.  Sorry about the neglect :)

**Comment by jayunit100** [comment]
Hmm what’s the test status now ? I can’t see the link to it.

**Pull request closed without merging**